kolena/annotation.py

```diff
@@ -382,6 +387,123 @@ class ScoredLabeledTimeSegment(TimeSegment):
     """The score associated with this time segment."""
 
 
+@dataclass(frozen=True, config=ValidatorConfig)
+class TextSegment(DataObject):
+    """
+    Segments of text in the associated text field
+    """
+
+    text_field: str
+    """Text field column name containing the text segments."""
+
+    segments: List[Tuple[int, int]]
+    """List of text segments. The first item of each tuple is the start index of the segment,
+    and the second item is the end index."""
+
+    def __post_init__(self) -> None:
+        if not self.segments:
+            raise ValueError("At least one text segment must be provided")
+        for start, end in self.segments:
+            if start < 0:
+                raise ValueError(f"Start index must be non-negative ({start} provided)")
+            if end < 0:
+                raise ValueError(f"End index must be non-negative ({end} provided)")
+            if start > end:
+                raise ValueError(f"Start index must be less than or equal to end index ({start} > {end})")
+
+
+def find_substring_indices(field_name: str, text: str, substrings: list[str]) -> Optional[TextSegment]:
+    indices = []
+    for substring in substrings:
+        escaped_search = re.escape(substring)
+
+        pattern = rf"\b{escaped_search}\b"
+
+        matches = re.finditer(pattern, text, re.IGNORECASE)
+
+        indices.extend([(match.start(), match.end()) for match in matches])
+
+    if indices:
+        return TextSegment(field_name, indices)
+    return None
+
+
+@dataclass(frozen=True, config=ValidatorConfig)
+class LabeledTextSegments(Annotation):
+    """
+    Segments of text in the associated text field with labels
+    ```
+    """
+
+    text_segments: List[TextSegment]
+    """List of text segments."""
+
+    label: str
+    """The label for the specified segments."""
+
+    color: Optional[str] = None
+    """Color of the segments for display."""
+
+    def __post_init__(self) -> None:
+        if not self.text_segments:
+            raise ValueError("At least one text segment must be provided")
+
+    @staticmethod
+    def _data_type() -> _AnnotationType:
+        return _AnnotationType.LabeledTextSegments
+
+    @staticmethod
+    def _create_from_keyword_single_row(
+        texts: dict[str, str],
+        keyword_labels: dict[str, list[str]],
+        colors: dict[str, str],
+    ) -> list["LabeledTextSegments"]:
+        labeled_segments = []
+        for label, keywords in keyword_labels.items():
+            segments = []
+            for field_name, text in texts.items():
+                segment = find_substring_indices(field_name, text, keywords)
+                if segment:
+                    segments.append(segment)
+            if segments:
+                labeled_segments.append(LabeledTextSegments(segments, label, colors.get(label, None)))
+        return labeled_segments
+
+    @staticmethod
+    def extract_labeled_text_segments_from_keywords(
```

Comment:
Odd to have this attached as a static method to this annotation class. Should this be defined elsewhere?

Reply:
Sure I can move it to util or something